ZcashFoundation/zcash-grant-system
# Rework Contribution Privacy

Closes #448.

![Screen Shot 2019-06-09 at 9 36 01 AM](https://user-images.githubusercontent.com/11430954/59160318-32991280-8a9a-11e9-8ad3-6e57134aa74b.png)

![Screen Shot 2019-06-09 at 9 36 55 AM](https://user-images.githubusercontent.com/11430954/59160321-388ef380-8a9a-11e9-8d58-55e0409e3e18.png)

### What this does
- authenticated users' contributions are never fully anonymous
- authenticated users can choose between private or public contributions
- private contributions will show up as "anonymous" on Proposal Detail view
- private contributions will be visible to to authenticated users on their Profile Funded tab, others will not see them 
- unauthenticated users' contributions will be fully anonymous
- unauthenticated users' contributions for cancelled proposals will become donations
- logged in users' contributions for cancelled proposals will be refundable if they have set their refund address
- `Proposal.no_refund` has been removed
- `Proposal.private` has been added
- migration: all Contributions associated with a user will be public (contributions made while unauthenticated, or authenticated but having selected the unattributed checkbox)

### Notes
- a "private" tag is used in the Profile > Funded view with a tooltip instead of the asterisk approach (see screenshot)

### Testing
1. clear your local DB & `flask db upgrade`
2. create two users
    - User A + set user's refund address (settings > account > refund address)
    - User B + no refund address
3. create, stake, approve + publish a proposal
4. **Anonymous** Sign-out, then contribute 0.1 ZEC to the proposal
    - after clicking "Fund this project" ensure that the disclaimer for unattributed contributions shows up
    - review verbiage
    - click "I accept"
    - payment info shows up, click "I've sent it", see "Thank you" message and click "Done"
3. **Refund Address/No Attribution** Sign-in with User A, then enter 0.2 ZEC into the proposal contribution field
    - ensure that "Contribute without attribution" is set by default, leave it selected
    - click "Fund this project" 
    - should see payment info, click "I've sent it", then "Done"
4. **No Refund Address/Attribution** Sign-in with User B, then enter 0.3 ZEC into contribution field
    - select "Attribute contribution publicly", click "Fund this project"
    - ensure Refund address setting modal appears, do not set a refund address
    - click "I don't want a refund, consider it a donation instead"
    - payment info should appear, click "I've sent it", then click "Done"
5. admin: Financials > Contributions should only have "staking 100% 0.025"
6. admin: Contributions > edit each of the pending contributions by setting the "Transaction ID" of each
7. admin: Financials > Contributions should have "funding 96% 0.6"
8. frontend: Proposal > Contributors (tab) should have three contributions and only one attributed to User B for 0.3 ZEC
9. frontend: Authenticate as User B and visit User A's Profile > Funded tab, there should be 0 contributions
10. frontend: Authenticate as User A and visit User A's Profile > Funded tab, there should be 1 private Contribution for 0.2 ZEC
    - visit User B's Profile > Funded tab, there should be 1 contribution for 0.3 ZEC
11. admin: from proposal detail "Cancel & refund" the proposal
12. admin: Financials > Contributions should have
    ```
    refunding 32% 0.2
    donation 64%  0.4
    staking   4%   0.025
    ```
13. admin: Home page should have one action item "ready to be refunded", click the action item link
    - there should be one Contribution from User A for 0.2 ZEC on the list page, click on the contribution
    - set a payment id (can be any string) and "Mark Refunded"
    - go back to Home, action item should be cleared
14. admin: Contributions > filter by "Donations", there should be two Contributions one from User B and the other from Anonymous
15. admin: Financials > Contributions should be:
    ```
    refunded 32% 0.2
    donation 64%  0.4
    staking   4%   0.025
    ```
